README.md
# SuperCAT
This is a README file to provide instructions on how to operate the SuperCat tool.
This tool currently supports Windows 7 and Window 10


[![Pester](https://github.com/lordneeko/SuperCAT/actions/workflows/PesterTest.yml/badge.svg)](https://github.com/lordneeko/SuperCAT/actions/workflows/PesterTest.yml)
[![PSScriptAnalyzer](https://github.com/SuperCATDevelopers/SuperCAT/actions/workflows/PSScriptAnalyzer.yml/badge.svg?branch=main)](https://github.com/SuperCATDevelopers/SuperCAT/actions/workflows/PSScriptAnalyzer.yml)

**********************************
*SuperCat Concept of Operation*
**********************************
The primary function of the SuperCAT tool is to scan offline systems to gather
metadata and audit logs on those systems.  In particular, SuperCAT has been
designed to execute solely from a CD/DVD disc so that the configuration baseline
of the scanned system is not affected. SuperCAT's command execute from the disc
and write back to the disc.  Although Windows will load the PowerShell scripts
(and executables) into memory while running, no files are expected to remain
resident on the system after the test completes.  Once the scan is complete, the
user should use the built in CD/DVD burner utilizies to write the added files
back to the disc and "finalize" the disc using a manner in accordance with
prescribed policies.

**********************************
*Preparation of the SuperCat Tool*
**********************************

After you have downloaded the SuperCat Tool you will need to populate the other
tools onto the disc to ensure you have the latest versions.

For the Anti-Virus Definition Files to be updated get the latest DAT file for
DOD Patch Repository under the ESS Collection.
https://patches.csd.disa.mil/CollectionInfo.aspx?id=863&bc=394_1_15_asc
* DAT files are places in the .\Tool\Scripts\McAfeeAV_v2\DAT\ folder

SuperCAT normally relies on the McAfee VirusScan Command Line scanner (VSCL) to operate.
* DOD Users may obtain a copy of this at the [DOD Patch Repository](https://patches.csd.disa.mil) by searching for VSCL Windows.
* Non-DOD, for information on obtaining this from McAfee is located here: https://kc.mcafee.com/corporate/index?page=content&id=KB5114.
* 64-bit VSCL is placed in the .\Tool\Scripts\McAfeeAV_v2\w64 folder.
* 32-bit VSCL is placed in the .\Tool\Scripts\McAfeeAV_v2\w32 folder.
** Note: Currently, a 32-bit scanner has not been identified. DOD Patch Repo no
longer provides the 32 bit scanner.

For the latest SCAP Compliance Checker (SCC) you can obtain that
tool on DoD Cyber Exchange.  https://cyber.mil/stigs/scap/
* Place SCC files within the .\Tool\Scripts\SCAP directory.

For users which want to scan older systems, you must download and install SCAP
content for those systems to the SCC by following the User's Manual provided
within the SCC download.  For those scanning current systems (e.g. Windows 10),
SCC already contains the latest SCAP content.

Once the scripts and files have been placed on the CD/DVD, you can complete the
burning process of the CD/DVD to take to the System(s) under test. Be sure when
adding the tool to the disc, that you do not "finalize" the session to ensure
that when the tests are executed Windows can add the files back to the disc.


****************************************
*****  Configuration :: Optional   *****
****************************************

NOTE:  This can be accomplished before getting to System(s) under test if you
have information already.

NOTE:  This step is NOT recommended when using one system per disk as it copies
drive IDs as well.

Configuration information from a pre-existing SuperCAT installation may be
duplicated to a new disk bycopying .\Tool\config.xml file to.\Tool\config.xml
on the new disk.



***********************************************
*****      Run SuperCAT :: Interactive    *****
***********************************************

NOTE:  Some items may need to be ran with admin credentials.  If this is the case
then when bringing up powershell, right click and select Run as Administrator.

Once you have optionally copied config.xml, you can proceed to run SuperCAT

Right click the SuperCAT script and run as Administrator to run.

The first menu to appear will allow you to select specific item to run or you
can run all of them.

After you have selected your option it will begin the process.  There will be a
prompt that tells you where it is in the process. It is recommended that the user
"finalize" the disc before attempting to eject the DVD.

**Note: Depending on the security requirements of the System Owner, you may
need to use a separate disc each time you scan a system to avoid the chance of
intermingling data or transferring malware.**

**********************************************************
*****      Run SuperCAT :: CommandLine :: Optional   *****
**********************************************************

NOTE: SuperCAT must have been run in interactive mode previously on this device.

In order to run SuperCAT in CommandLine mode:
 1. Open up Windows Explore
 2. Select the SuperCAT disk
 3. Navigate to the Tools folder
 4. Click File
 5. Hover over the arrow by Open Powershell
 6. Select Open Powershell as Administrator.

.\SuperCAT.ps1 [-Help] [-List]
.\SuperCAT.ps1 [-Help] [-List] -Time TIME -Options OPTION[,Option...]

Parameters:
  -Help     Display this message.
  -List     Display only execution options
  -Options  Select one or more execution options. Please
            only input numbers and commas (i.e. 1,25,6).
            Requires -Time
  -Time     Set the system time in UTC. To skip, write
            "trust". Please enter the date in the format
            YYYY-MM-DDTHH:MM:SS (i.e. 2020-01-31T13:39:00)

Execution Options:
	0 = Update Antivirus (Requires Pre-Approved Actions)
	1 = Collect Computer Information
	2 = Initialize Antivirus Scan
	3 = Collect Antivirus Logs
	4 = Initialize SCAP
	5 = Collect Windows Event Logs
	6 = All Tasks (No Antivirus Updating, Auto exits)
	7 = Exit Program

************************************
***    Results and Info       ******
************************************

Once it is completed the CD/DVD will contain files with the extracted information
for each system within the Outputs folder.


************************************
***   Event ID Analysis          ***
************************************

SuperCAT gets an extract of Windows Event logs and stores them as raw EVTX files.
These can be injested directly with logstash. In addition, within the
Tools\Scripts\QuicklookTools folder is a tool (FilterEventLogs.ps1) for parsing
the evtx files and converting them to csv files.  The user can supply a txt file
of EventIDs for which to filter the CSV and it will output an analysis file with
the EventIDs the user is interested in. The user can get a good list of EventIDs
to use for Incident Response, Threat Hunting, Forensics, etc at the following location:
https://github.com/TonyPhipps/SIEM/blob/ec2dde7ba7997bfb9a88acf27fbb1fde7e32d20c/Notable-Event-IDs.md


ReleaseNotes.txt
v2.00
- Rewrote the original batch script into PowerShell.

v2.01
- Added PowerShell calling for a .JSON file for configurations.
- Added an additional external script to create the JSON file.
- Added an additional external script to query .EVTX files.

v2.10
- Fixed an issue where several values were not being added to a file.
- Added check to remove files added to the local computer's temporary folder.
- Added check to see if script was ran as administrator.
- Added new prompt that will ask if operator would like to remove extra files and folders after the collection is completed.
- Added "Exit Without Collection" option.
- Moved Antivirus Updating to Option 0, moving all other options up.

v2.11
- Removed code that would create a new temporary folder on the operated machine, keeping in line the passive capabilities of the script.

v2.20
- Added hard drive information (including serial number and model) to gathering task.
- Added MAC address collection to gathering tasks; will only pull the very first MAC.
- Fixed an issue where setup-powershell.ps1 does not write base name to JSON file.
- Fixed an issue where setup-powershell.ps1 was not overwriting the JSON file properly.
- Fixed an issue where Move-Item commands mistakenly had -Recurse as a parameter.
- Changed/simplified save location of exported/collected files; will now appear above the directory above where the script is ran.
- Changed the $ComputerName_$SerialNumber.txt file holding basic user input to $ComputerName-Info.txt
- Added EVTX QuickLook Analysis tool

=======

v0.3.0
- BREAKING CHANGE Switched configuration file to XML for better object handling.
- BREAKING CHANGE Regressed #42 Commandline arguments; pending reimplementation.
- Rewrote script to utilize functions and modules.
- Moved configuration generation into SuperCAT.ps1.
- Added UI framework.
- Implemented #50 .gitignore for config.xml
- Provided for accurate-ish datetime on systems with dead a CMOS battery.
- Moved towards compliance with Semantic Versioning 2.0.0 https://semver.org/
  - Incremented to v0.3.0 due to API change.
  - Selected major version 0 in move to Semantic Versioning due to lack of stable API. See https://semver.org/#spec-item-4

v0.3.1
- Implemented issue #52.
- Switched to changing system time from getting time delta.
- Restructured file system (again).
- Implemented common naming scheme.
  - yyyyMMdd_Scan#_MXUnit_SystemAbreviation+Version_DriveID_Filename
  - 20211230_1_480AMXS_CA10_001_Info.txt
- Incremented config.xml to v2.1.0.
  - Tracked SystemVersion in config.xml

v0.3.2
- Fixed .gitignore.
- Implemented hold for AV and SCAP completion.
- Implemented duplication check in UI.
- Accounted for running at midnight in filename generation.
- Reduced confusion in menu by removing used options.
  - Reimplemented main loop as for instead of while
- Removed newlines for consistency.

v0.3.3
- Added check for config.xml version.
- Incremented config.xml version: v2.2.0.
  - Removed SystemVersion from config.xml.
- Added newlines for clarity.
- Removed unnecessary verbiage.
- Switched from removing entries to marking completed in menu.
- Added code comments for clarity.
- Reimplemented main loop as while instead of for due to PS abnormalities.

v0.3.4
- Changed config.xml version to fit Semantic Versioning, v0.2.2.
- Updated ReleaseNotes.txt.
- Removed development environment artifact.
- Implemented issue #43, write out drive serial number.
- Changed function nouns to singular iaw PS style guide.



tests/Structure.Tests.ps1
# This test validates the folder structure of SuperCAT to ensure that all the essential files will be in the correct location on disk.
Describe 'SuperCAT Structure' {
	Context 'File Structure' {
		It 'DISA Directory Check' {
			$currentDir = [string](Get-Location)
			$currentDir + "\Tool\Scripts\SCAP" | Should -Exist
		}
		It 'AV Directory Check' {
			$currentDir = [string](Get-Location)
			$currentDir + "\Tool\Scripts\McAfeeAV_v2" | Should -Exist
			$currentDir + "\Tool\Scripts\McAfeeAV_v2\w32" | Should -Exist
			$currentDir + "\Tool\Scripts\McAfeeAV_v2\w64" | Should -Exist
			$currentDir + "\Tool\Scripts\McAfeeAV_v2\DAT" | Should -Exist
		}
		It 'QuickLookTool checks' {
			$currentDir = [string](Get-Location)
			$currentDir + "\Tool\Scripts\QuickLookTools\FilterEventLogs.ps1" | Should -Exist
			$currentDir + "\Tool\Scripts\QuickLookTools\export-offline-registry.ps1" | Should -Exist
		}
		It 'SuperCAT main file checks' {
			$currentDir = [string](Get-Location)
			$currentDir + "\Tool\SuperCAT.ps1" | Should -Exist
			$currentDir + "\Tool\Modules\Support.psm1" | Should -Exist
		}
	}
}



setup-powershell.ps1
Clear-Host
Write-Output "
===============================================================
This script helps automate the creation of the config.json file
used by the Cyber Assessment Tool (CAT). Whenever changes are
needed for the file, you can either edit the variables within
the file or you can use this script again to create it.
==============================================================="
pause

$configFileExists = Test-Path -Path Tool/config.json -PathType Leaf

If($configFileExists) {
    Write-Output "
    ===============================================================
    Configuration already exists at .Tool\config.json
    Continuing will overwrite the current configuration.
    Do you want to continue?
    1 = Yes
    2 = No
    ==============================================================="
    $overwriteConfig = Read-Host

    while($overwriteConfig -notmatch "^[1-2]$"){
        $overwriteConfig = Read-Host "You must input either 1 or 2."
    }

    if($overwriteConfig -eq 2){
        # Cleans up the disc of files no longer required.
        Write-Output "Exiting..."
        exit
    }
}




while(!($Basecheck -eq 1)){
    $Base = Read-Host "Type in the name of your base"

    Write-Output "
    ===============================================================
    You inputted $Base . Is this correct?
    1 = Yes
    2 = No
    ==============================================================="
    $BaseCheck = Read-Host
}

while(!($SystemCheck -eq 1)){
    Clear-Host
    $System1 = Read-Host "Type in the first common system you will be assessing"
    $System2 = Read-Host "Type in the second common system you will be assessing"
    $System3 = Read-Host "Type in the third common system you will be assessing"
    $System4 = Read-Host "Type in the fourth common system you will be assessing"
    $System5 = Read-Host "Type in the fifth common system you will be assessing"

    Write-Output "
    ===============================================================
    You inputted:
    $System1
    $System2
    $System3
    $System4
    $System5
    1 = Yes
    2 = No
    ==============================================================="
    $SystemCheck = Read-Host
}

while(!($LocationCheck -eq 1)){
    Clear-Host
    $Location1 = Read-Host "Type in the first common location you will be assessing"
    $Location2 = Read-Host "Type in the second common location you will be assessing"
    $Location3 = Read-Host "Type in the third common location you will be assessing"
    $Location4 = Read-Host "Type in the fourth common location you will be assessing"
    $Location5 = Read-Host "Type in the fifth common location you will be assessing"

    Write-Output "
    ===============================================================
    You inputted:
    $Location1
    $Location2
    $Location3
    $Location4
    $Location5
    1 = Yes
    2 = No
    ==============================================================="
    $LocationCheck = Read-Host
}


[ordered]@{
    BaseName = $Base
    System1 = $System1
    System2 = $System2
    System3 = $System3
    System4 = $System4
    System5 = $System5
    Location1 = $Location1
    Location2 = $Location2
    Location3 = $Location3
    Location4 = $Location4
    Location5 = $Location5
} | ConvertTo-JSON | Set-Content -Path "Tool/config.json"

Get-Content Tool/config.json

Write-Output "
===============================================================
config.json has been created!
==============================================================="

Export/export-offline-eventlogs.ps1
$Drive = Read-Host "Type in the exact path the .evtx event files are located"
$EventLogs = (Get-ChildItem -Path "$Drive" | Where-Object {$_.Extension -eq ".evtx"}).fullname

$EventIDsFile = Read-Host "Type in the exact path the whitespace-delimited text file of event IDs"
$EventIDs = (Get-Content $EventIDsFile).Trim().Split()

foreach($EventLog in $EventLogs){
    $Type = (Get-ChildItem $EventLog).BaseName
    Write-Output "Writing $Type to CSV... This may take some time..."
    Get-WinEvent -Path $EventLog | Sort-Object -Property TimeCreated -Descending | Export-CSV -NoTypeInformation -Path "$Drive\EXPORTED-$Type-EVTX.csv"
	$CSV = Import-Csv "$Drive\EXPORTED-$Type-EVTX.csv"
	$row = 0
	foreach ($line in $CSV){
		$row++
		if($EventIDs -contains $line.Id){
			Write-Output (-join("EventID ", $line.Id, " found in EXPORTED-$Type-EVTX.csv on line $row"))
			Write-Output $line | Export-Csv -Path "$Drive\FLAGGED-EVENTS.csv" -Append -NoTypeInformation
		}
	}
}

Write-Output "Flagged Events written to $Drive\FLAGGED-EVENTS.csv"
Write-Output "Job's done!"

Tool/SuperCAT.ps1
<###############################################################################
## SUPERCAT (CYBER ASSESSMENT TOOL) V0.3.4
## DEVELOPED BY: SSGT CLINTON REEL // CLINTON.REEL@US.AF.MIL
## ADDITIONAL DEVELOPERS IN CONTRIBUTORS.TXT
###############################################################################>

#########################  Parameters  #########################################
[CmdletBinding(DefaultParameterSetName="Interactive")]
param (
    [Alias("h","?")]
    [Parameter(ParameterSetName="Cmdline")]
    [Parameter(ParameterSetName="Interactive")]
    [Switch]$Help,

    [Parameter(ParameterSetName="Cmdline")]
    [Parameter(ParameterSetName="Interactive")]
    [Switch]$List,

    [Parameter(Mandatory=$True,Position=0,ParameterSetName="Cmdline")]
    [String]$Options,

    [Parameter(ParameterSetName="Interactive",
    HelpMessage="Please enter the UTC date and time in the format YYYY-MM-DDTHH:MM:SS (i.e. 2020-01-01T13:39:00)")]
    [Parameter(Mandatory=$True,Position=1,ParameterSetName="Cmdline",
    HelpMessage="Please enter the UTC date and time in the format YYYY-MM-DDTHH:MM:SS (i.e. 2020-01-01T13:39:00)")]
    [String]$Time
)

##################### Early Variable Requirements ##############################

$ScriptVersion = [System.Version]"0.3.4"

$ScriptDirectory = $MyInvocation.MyCommand.Path | Split-Path

$AllOptions = @(
    "Update Antivirus (Requires Pre-Approved Actions)",
    "Collect Computer Information",
    "Initialize Antivirus Scan",
    "Collect Antivirus Logs",
    "Initialize SCAP",
    "Collect Windows Event Logs",
    "All Tasks (No Antivirus Updating, Auto exits)",
    "Exit Program"
)
$AdminRequired = @(
    "Update Antivirus (Requires Pre-Approved Actions)",
    "Initialize Antivirus Scan",
    "Initialize SCAP",
    "All Tasks (No Antivirus Updating, Auto exits)"
)

$RemainingOptions = @()
$RemainingOptions += $AllOptions

############################# Module Import ####################################
try {
    Import-Module -Force -ErrorAction Stop "$ScriptDirectory\Modules\Support.psm1"
} catch {
    throw "Missing Support.psm1!"
}

try {Import-Module -Force -ErrorAction Stop "$ScriptDirectory\Modules\Antivirus.psm1"}
catch {
    Write-Warning "Missing Antivirus.psm1, unable to run, update, or import logs from antivirus."
    $RemainingOptions[0] = $RemainingOptions[0].Insert(0,"(Unavailable) ")
    $RemainingOptions[2] = $RemainingOptions[2].Insert(0,"(Unavailable) ")
    $RemainingOptions[3] = $RemainingOptions[3].Insert(0,"(Unavailable) ")
}
try {Import-Module -Force -ErrorAction Stop "$ScriptDirectory\Modules\EventLogs.psm1"}
catch {
    Write-Warning "Missing EventLogs.psm1, unable to import event logs."
    $RemainingOptions[5] = $RemainingOptions[5].Insert(0,"(Unavailable) ")
}
try {Import-Module -Force -ErrorAction Stop "$ScriptDirectory\Modules\GeneralCollection.psm1"}
catch {
    Write-Warning "Missing GeneralCollection.psm1, unable to collect general information."
    $RemainingOptions[1] = $RemainingOptions[1].Insert(0,"(Unavailable) ")
}
try {Import-Module -Force -ErrorAction Stop "$ScriptDirectory\Modules\SCAP.psm1"}
catch {
    Write-Warning "Missing SCAP.psm1, unable to run SCAP."
    $RemainingOptions[4] = $RemainingOptions[4].Insert(0,"(Unavailable) ")
}

## Mark "All Tasks" unavailable if any of the functions are unavailable.
if ($RemainingOptions[1..$($RemainingOptions.GetUpperBound(0)-2)] -contains "(Unavailable) ") {
    $RemainingOptions[-2] = $RemainingOptions[-2].Insert(0,"(Unavailable) ")
}

##################### Parameter Processing #####################################

if ($Help) {
    Write-Host
    Write-Host "SuperCAT version $($($ScriptVersion).ToString())"
    Write-Host
    Write-Host "System Requirements:"
    Write-Host "Windows 7 or greater with Powershell 2 or greater."
    Write-Host
    Write-Host ".\SuperCAT.ps1 [-Help] [-List] [-Time TIME]"
    Write-Host ".\SuperCAT.ps1 [-Help] [-List] -Time TIME -Options OPTION[,Option...]"
    Write-Host
    Write-Host "Parameters:"
    Write-Host "  -Help     Display this message."
    Write-Host "  -List     Display only execution options"
    Write-Host "  -Options  Select one or more execution options. Please"
    Write-Host "            only input numbers and commas (i.e. 1,25,6)."
    Write-Host "            Requires -Time"
    Write-Host "  -Time     Set the system time in UTC. To skip, write"
    Write-Host "            `"trust`". Please enter the date in the format"
    Write-Host "            YYYY-MM-DDTHH:MM:SS (i.e. 2020-01-31T13:39:00)"
    Write-Host "                      ^ Notice the T"
    Write-Host
    Write-Host "Execution Options:"
    for ($i=0; $i -lt $AllOptions.Count; $i++) {
        Write-Host $i "=" $AllOptions[$i]
    }
    Write-Host
    exit
}
if ($List) {
    Write-Host
    Write-Host "The following are your options, please"
    Write-Host "input only numbers and commas (i.e. 1,25,6):"
    for ($i=0; $i -lt $AllOptions.Count; $i++) {
        Write-Host $i "=" $AllOptions[$i]
    }
    Write-Host
    exit
}
if ( $Options ) {
    $ResultList = $Options.Split(",")
    if ($Options -notmatch "^\d+(,\d+)*$") {
        throw "Please only input numbers and commas (i.e. 1,25,6)"
    }
    elseif (( [int[]]$ResultList -ge $AllOptions.Count ) -or ( $ResultList -lt 0 )) {
        throw "Please ensure your entry is between 0 and $($AllOptions.Count - 1)"
    }
    elseif ($(Get-Duplicate $ResultList)) {
        throw "Please ensure there are no duplicates in your entry."
    }
    else {
        $Chosen = $AllOptions[$ResultList]
    }
    $NoInteractive=$True
} else { $NoInteractive=$False }

if (Test-Path -Path "$ScriptDirectory\SuperCAT.lck") {
    Write-Host "SuperCAT already running. If you believe this is an error,"
    Write-Host "delete SuperCAT.lck"
    exit
} else {
    New-Item -ItemType file "$ScriptDirectory\SuperCAT.lck" | Out-Null
}


######################### Configuration Handling ###############################

function Import-Config {
    # .SYNOPSIS
    # Pull config object from file. Generates skeleton if necessary.
    param (
        [Parameter(Mandatory=$True,ValueFromPipeline=$True,Position=0)]
        [ValidateNotNullOrEmpty()]
        [String]$File,

        [Parameter()]
        [Bool]$NoInteractive
    )
    $ConfigVersion = [System.Version]"0.2.3"
    if (Test-Path -Path $File -PathType Leaf) {
        Try {
            $Config = $(Import-Clixml -ErrorAction Stop -Path $File)
        }
        Catch {
            throw "The file $File incorrectly formated, please delete to force recreation."
        }
    }
    elseif ( Test-Path -Path $File -PathType Container ) {
        throw "$File is a directory!"
    }
    elseif ( $NoInteractive ) {
        throw "Please run interactively to generate config.xml first."
    }
    else {
        $Config = [PSCustomObject]@{
            ConfigVersion       = $ConfigVersion
            CreationTimeUTC     = Get-Date
            LastAccessTimeUTC   = Get-Date
            LastWriteTimeUTC    = Get-Date
            LastHDD             = ''
            Location            = Read-Host -Prompt "Which location is this"
            ScanningOrg         = Read-Host -Prompt "What is the scanning organization"
            KnownDrives         = @{}
        }
    }
    if ($Config.ConfigVersion -ne $ConfigVersion) { throw "Old config file, failing out."}
    Write-Host
    return $Config
}
function Update-Config {
    # .SYNOPSIS
    # Update config object, adding drive config if it doesn't exist.
    param (
        [Parameter(Mandatory=$True,ValueFromPipeline=$True,Position=0)]
        [PSCustomObject]$Config,

        [Parameter(Mandatory=$True,Position=1)]
        [ValidateNotNullOrEmpty()]
        [String]$RootDirectory,

        [Parameter()]
        [Bool]$NoInteractive
    )
    function Read-DriveName {
        # .SYNOPSIS
        # Ask the user for the drive number, padding to 3 digits.
        Write-Host "================================================"
        Write-Host "Please enter the assigned drive number, between"
        Write-Host "0 and 999. Example: 27"
        Write-Host "================================================"
        while ($True) {
            try {
                [int]$Result = Read-Host -Prompt "Drive Number"
            } catch {
                Write-Host "Please enter a number."
            }
            if ($Result -gt 999) {
                Write-Host "Please enter a number less than 999."
            }
            else {
                return $([String]$Result).PadLeft(3,"0")
            }
        }
    }

    $LocalDrive = $(Get-WmiObject Win32_PhysicalMedia |
        Where-Object {$_.Dependent -eq $(
        Get-WmiObject Win32_DiskDriveToDiskPartition |
        Where-Object {$_.Dependent -eq $(
        Get-WmiObject Win32_LogicalDiskToPartition |
        Where-Object {$_.Dependent -Like $Env:SystemDrive})})}).SerialNumber
    $Config.LastAccessTimeUTC = Get-Date
    if ($Config.KnownDrives.Keys -NotContains $LocalDrive) {
        if ($NoInteractive) {
            throw "Please run interactively first to generate config.xml"
        }
        if ($Config.KnownDrives.Count -gt 0) {
            Write-Host "================================================"
            Write-Host "Unknown drive. Would you like to add this `ndrive to database?"
            Write-Host "================================================"
            if ( !$(Read-Intent -TF) ) {
                Write-Host "Exiting! Nothing has been written."
                if (Test-Path -Path "$ScriptDirectory\SuperCAT.lck") {
                    Remove-Item -Path "$ScriptDirectory\SuperCAT.lck"
                }
                exit
            }
            else {
                Write-Host "Adding drive to database."
                Write-Host
            }
        }
        $Config.KnownDrives.$LocalDrive = @{
            CreationTimeUTC     = Get-Date
            LastAccessTimeUTC   = Get-Date
            LastWriteTimeUTC    = Get-Date
            DriveName           = Read-DriveName
            SystemType          = Read-CSV "$RootDirectory\SystemList.csv" "SystemName"
            SystemOwner         = Read-Host -Prompt "What organization does this system belong to"
            Classification      = Read-CSV "$RootDirectory\ClassificationList.csv" "Classification"
        }
    }
    while ($True) {
        Write-Host
        Write-Host "================================================"
        if (!$NoInteractive) {
            Write-Host "Is this information correct?"
            Write-Host
        }
        Write-Host "Location            = $($Config.Location)"
        Write-Host "Organization        = $($Config.ScanningOrg)"
        Write-Host "Drive Name          = $($Config.KnownDrives.$LocalDrive.DriveName)"
        Write-Host "System Type         = $($Config.KnownDrives.$LocalDrive.SystemType)"
        Write-Host "System Owner        = $($Config.KnownDrives.$LocalDrive.SystemOwner)"
        Write-Host "Classification      = $($Config.KnownDrives.$LocalDrive.Classification)"
        Write-Host "================================================"
        if ($NoInteractive -or $(Read-Intent -TF)) { Write-Host; break }
        $Config.KnownDrives.$LocalDrive.LastWriteTimeUTC = Get-Date
        $Config.LastWriteTimeUTC = Get-Date
        $SelectionArray = @(
            "Location",
            "Organization",
            "Drive Name",
            "System Type",
            "System Owner",
            "Classification"
        )
        switch($(Read-Intent $SelectionArray "What should be changed?")) {
            $SelectionArray[0] {$Config.Location                                = Read-Host -Prompt "Location"}
            $SelectionArray[1] {$Config.ScanningOrg                             = Read-Host -Prompt "Organization"}
            $SelectionArray[2] {$Config.KnownDrives.$LocalDrive.DriveName       = Read-DriveName}
            $SelectionArray[3] {$Config.KnownDrives.$LocalDrive.SystemType      = Read-CSV "$RootDirectory\SystemList.csv" "SystemName"}
            $SelectionArray[4] {$Config.KnownDrives.$LocalDrive.SystemOwner     = Read-Host -Prompt "System Owner"}
            $SelectionArray[5] {$Config.KnownDrives.$LocalDrive.Classification  = Read-CSV "$RootDirectory\ClassificationList.csv" "Classification"}
            Default {throw "Update-Config switch fell through."}
        }

    }
    $Config.KnownDrives.$LocalDrive.LastAccessTimeUTC = Get-Date
    $Config.LastHDD = $LocalDrive
    return $Config
}
function Export-Config {
    #.SYNOPSIS
    # Write config object to file.
    param (
        [Parameter(Mandatory=$True,ValueFromPipeline=$True,Position=1)]
        [PSCustomObject]$Config,

        [Parameter(Mandatory=$True,Position=0)]
        [String]$File,

        [Parameter()]
        [Switch]$PassThru
    )
    #$Config | Select-Object -ExcludeProperty 'LastHDD' -Property * |
    #  Export-Clixml -Depth 10 -Force -Path $File | Out-Null
    $Config | Export-Clixml -Depth 10 -Force -Path $File | Out-Null
    if ($PassThru) {
        return $Config
    }
    else{
        return $True
    }
}
function Get-LogPrefix {
    # .SYNOPSIS
    # Pull system abbreviations and mesh config data to provide a log prefix.
    param (
        [Parameter(Mandatory=$True,ValueFromPipeline=$True,Position=0)]
        [PSCustomObject]$Config,

        [Parameter(Mandatory=$True,Position=1)]
        [ValidateNotNullOrEmpty()]
        [String]$RootDirectory,

        [Parameter()]
        [Switch]$SCAP
    )
    $ClassAbbreviation = Read-CSV `
        -Path   "$RootDirectory\ClassificationList.csv" `
        -Column "Classification" `
        -Select "ClassificationAbbreviation" `
        -Config $Config
    $SystemAbbreviation =Read-CSV `
        -Path   "$RootDirectory\SystemList.csv" `
        -Column "SystemName" `
        -Select "SystemAbbreviation" `
        -Config $Config
    $SystemOwner  = $Config.KnownDrives.$($Config.LastHDD).SystemOwner
    $Date         = $(Get-Date -Format "yyyyMMdd_HHmm" -Date $Config.LastAccessTimeUTC)
    $Drive        = $Config.KnownDrives.$($Config.LastHDD).DriveName
    if ($SCAP) {
        return "$ClassAbbreviation`_$SystemOwner`_$SystemAbbreviation`_$Drive`_$($Config.LastHDD)"
    } else {
        return "$ClassAbbreviation`_$Date`_$SystemOwner`_$SystemAbbreviation`_$Drive`_$($Config.LastHDD)"
    }
}

################################ General #######################################

## Check for administrative privildges, warn if necessary, continue
if(([Security.Principal.WindowsPrincipal](
    [Security.Principal.WindowsIdentity]::GetCurrent())).IsInRole(
    [Security.Principal.WindowsBuiltInRole]::Administrator) -eq $False){
    Write-Warning "THIS SCRIPT WAS NOT RUN AS AN ADMINISTRATOR!"
    Write-Warning "SOME TASKS ARE UNAVAILABLE OR MAY PROVIDE INCOMPLETE RESULTS!"
    foreach ($Item in $AdminRequired) {
        $Index = $RemainingOptions.IndexOf($Item)
        if ($RemainingOptions[$Index] -notcontains "(Unavailable) ") {
            $RemainingOptions[$Index] = $RemainingOptions[$Index].Insert(0,"(Unavailable) ")
        }
    }
}
$LogPath = "$ScriptDirectory..\..\Outputs"
## Fix time on the laptop, and set it to UTC
Set-Time $Time | Out-Null

$Config = Import-Config "$ScriptDirectory\config.xml" -NoInteractive $NoInteractive |
    Update-Config -RootDirectory $ScriptDirectory -NoInteractive $NoInteractive |
    Export-Config "$ScriptDirectory\config.xml" -PassThru
$LogPrefix = Get-LogPrefix $Config $ScriptDirectory
$ExitLock = @() ## Keep track of what programs have been started.
while (($Chosen -notcontains $AllOptions[-1]) -or ($Options)) {
    if (!($Options)) {
        $Chosen = Read-Intent $RemainingOptions -Multiple
    }
    ## Handles "All Tasks" by setting $Chosen to everything but AV Update.
    ## Unless of course the user selects AV Update as well.
    if ($Chosen -contains $AllOptions[-2]) {
        if ($Chosen -contains $AllOptions[0]) { ## AV
            $Chosen = $AllOptions[0..$($AllOptions.GetUpperBound(0)-2)]
        }
        else { ## No AV
            $Chosen = $AllOptions[1..$($AllOptions.GetUpperBound(0)-2)]
        }
        $Chosen += $AllOptions[-1]
    }
    switch($Chosen) {
        $AllOptions[0]  { Update-AVSignature $ScriptDirectory }
        $AllOptions[1]  { Import-Identifier $Config "$LogPath\GatheredLogs\$LogPrefix" }
        $AllOptions[2]  { $ExitLock += Start-Antivirus $ScriptDirectory "$LogPath\McAfeeLogs\$LogPrefix" }
        $AllOptions[3]  { Import-AntivirusLog "$LogPath\McAfeeLogs\$LogPrefix" "$LogPath\EventLogs\$LogPrefix" }
        $AllOptions[4]  { $ExitLock += Start-SCAP $ScriptDirectory "$LogPath\SCAPLogs\$(Get-LogPrefix $Config $ScriptDirectory -SCAP)" }
        $AllOptions[5]  { Import-EventLog "$LogPath\EventLogs\$LogPrefix"}
        $AllOptions[-2] { throw "`$AllOptions[-2] if statement failed to evaluate correctly." }
        $AllOptions[-1] { Out-Null }
        {$_.Contains("(Complete) ")} {
            Write-Host "Skipping $($_.Remove(0, 11)); already run"
        }
        {$_.Contains("(Unavailable) ")} {
            Write-Host "Skipping $($_.Remove(0, 14)); some elements run, module not imported, or running as user."
        }
        Default {
            Write-Host "Please only input numbers 0 to $($AllOptions.GetUpperBound(0)) and commas (i.e. 1,3,5)"
            Write-Host "Recieved Input $_"
        }
    }
    ## Mark all selected options as complete in the list, preventing their execution.
    foreach ($Selected in $Chosen.Where({($_ -ne $AllOptions[-1]) -and
        ($_ -notcontains "(Complete)") -and ($_ -notcontains "(Unavailable)")})) {
        $Index = $RemainingOptions.IndexOf($Selected)
        $RemainingOptions[$Index] = $RemainingOptions[$Index].Insert(0,"(Complete) ")

        if (($RemainingOptions[-2] -eq $AllOptions[-2]) -and
            ($AllOptions[1..$($AllOptions.GetUpperBound(0)-2)] -contains $Selected)) {
            $RemainingOptions[-2] = $RemainingOptions[-2].Insert(0,"(Unavailable) ")
        }
    }
    Write-Host
    if ( $Options ) { break }
}

## Prevent exiting if ExitLock still has an active process
if (($ExitLock.Count -ne 0) -and ($ExitLock.HasExited -contains $False)) {
    Write-Host "The following processes are still running: $($ExitLock.Where(
        {$_.HasExited -eq $False}).ProcessName)"
    Write-Host "Waiting for processes to finish up..."
    $ExitLock.WaitForExit()
    Write-Host "Done!"
}

if (Test-Path -Path "$ScriptDirectory\SuperCAT.lck") {
    Remove-Item -Path "$ScriptDirectory\SuperCAT.lck"
}
